From 8d7123a40c66292060fc7ee1bc07e7c61c3a2dd7 Mon Sep 17 00:00:00 2001
From: Sergei Zimmerman <sergei@zimmerman.foo>
Date: Fri, 3 Apr 2026 00:21:21 +0300
Subject: [PATCH 1/2] libstore: Make temporary in-store directory not
 world-readable

(cherry picked from commit 0e3412a93f43a017342c267000c152e5c45327e7)
(cherry picked from commit a3215e7c5c260fab5f2cb034c4df01dfa3b284e5)
---
 src/libstore/local-store.cc | 5 +++--
 1 file changed, 3 insertions(+), 2 deletions(-)

diff --git a/src/libstore/local-store.cc b/src/libstore/local-store.cc
index c5489444e..c82c1f3ad 100644
--- a/src/libstore/local-store.cc
+++ b/src/libstore/local-store.cc
@@ -1375,8 +1375,9 @@ std::pair<std::filesystem::path, AutoCloseFD> LocalStore::createTempDirInStore()
     do {
         /* There is a slight possibility that `tmpDir' gets deleted by
            the GC between createTempDir() and when we acquire a lock on it.
-           We'll repeat until 'tmpDir' exists and we've locked it. */
-        tmpDirFn = createTempDir(realStoreDir, "tmp");
+           We'll repeat until 'tmpDir' exists and we've locked it.
+           Make the directory accessible only to the current user.*/
+        tmpDirFn = createTempDir(realStoreDir, "tmp", /*mode=*/0700);
         tmpDirFd = openDirectory(tmpDirFn);
         if (!tmpDirFd) {
             continue;
-- 
2.53.0

